### PR TITLE
[RFC WIP] New ctfe

### PR DESCRIPTION
**Pull request by UplinkCoder**
I open this PR just so I can see comments better. 
I will adhere to Dstyle as soon as this is in a more complete state.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
@MartinNowak  The reason why I have a `short`-jump is so I can jump forwards and backwards by adding the signed offset to the current Instruction-pointer.

**Comment by UplinkCoder** [comment]
I will split it into modules and do the rest of the cleanup when I can interpret bf-ctfe or  when I get fed up with the state of the source whicherver happens first.
NOTE: The source is still in flux and is expected the change a lot in response to observed performance concerns or suggestions/comments.

**Comment by UplinkCoder** [comment]
Opps I just got what you were getting at with your short instruction coding comment.
The answer is simple. 16 byte instructions don't pay for themselves. Having another flag to check for slowes down the interpreter loop.

**Comment by UplinkCoder** [comment]
@MartinNowak I did a few performance tests.
So far the byteCode-generation takes about as much time as the error check in ctfeExpr.

**Comment by UplinkCoder** [comment]
awww dammit

**Comment by MartinNowak** [comment]
I'd rather like to integrate this into a feature branch (e.g. dlang/dmd#ctfe) using multiple PRs until it can fully replace the existing implementation.

**Comment by MartinNowak** [comment]
Your git history is a mess ;), but we can deal with that later.

**Comment by UplinkCoder** [comment]
If my engine just triggers in 80% of the "hot-loop cases" where OOM was a problem before,  it removes a blocker to d's adoption instantly.
The point here is, most bang(!) for the buck. 

If your concern if having a stable release, I can assure you that merging a feature complete but not field-tested engine is way more trouble then a half-finished but malleable engine. 
This is not is not about a language issue. 
It is a performance optimization! 
Even if it just covers 10% of the cases as long as the cases are hot cases it is a success.

**Comment by UplinkCoder** [comment]
I just cleaned it a little bit.
@MartinNowak your comments were very helpful.

**Comment by UplinkCoder** [comment]
progress is currently blocked by PR #5920

File: src/bc.d [hunk 1]
```diff
@@ -31,7 +31,7 @@ enum ShortInst : ubyte
     Neg,
     Soh, /// set hi stack addr
     Sdh, /// set hi dataseg addr
-
+    Flg, // writes the conditionFlag into [lw >> 16]
```
Comment (UplinkCoder) [inline src/bc.d]: not really a need. 
I just like them because that is what asm looks like.
If you have better names please suggest them.

**Comment by UplinkCoder** [comment]
Thanks! @klickverbot

**Comment by UplinkCoder** [comment]
It should be green now. 
Please be advised this is not because the engine passes the tests but merely because I deactivated it by default.

**Comment by ZombineDev** [comment]
OT: @UplinkCoder I noticed that you have lots of "fix ws" commits. What editor are you using?

**Comment by UplinkCoder** [comment]
@ZombineDev notepad++, On Windows,
nano on linux.
Both do not strip trailing Whitespace in the current configuration at least.

**Comment by ZombineDev** [comment]
@UplinkCoder 

Many projects like [dmd](https://github.com/dlang/dmd/blob/master/.editorconfig), [druntime](https://github.com/dlang/druntime/blob/master/.editorconfig) and [phobos](https://github.com/dlang/phobos/blob/master/.editorconfig) use [EditorConfig](http://editorconfig.org/) to automatically enforce whitespace style. There plugins for most major editors and IDEs which makes it easy to use.

If you install the [EditorConfig plugin for Notepad++](https://github.com/editorconfig/editorconfig-notepad-plus-plus) it should take care of automatically fixing the whitespace problems you have. You don't need to do any manual configuration - the editor plugin should detect the `.editorconfig` file and automatically configure the editor for you everytime you open a file in a directory which contains an `.editorconfig` file (or if it's parent directory contains it - recursively).

On linux I usually use vim and sublime-text (which both have editorconfig plugins), so I'm not sure about nano. What I could find though was that you can put the following in `~/.nanorc`:
```
set tabsize 4
set tabstospaces
```

Unfortunately, I couldn't find an option to automatically trim whitespace, but you could use `Alt+P` to make the editor display it for you.

**Comment by UplinkCoder** [comment]
@ZombineDev thanks for doing the research :+1: 
I appreciate it.

**Pull request closed without merging**